Added FreeRTOS

LHelge opened the pull request:
Played around with FreeRTOS and created a c++ wrapper for tasks. The plan is to put other stuff, like the button, accelerometer and debug in their own tasks. Perhaps wrap some more parts of FreeRTOS like Mutex and Queues to pass events to the state machine.

Any thoughts? @abbeman

LHelge committed:
Added FreeRTOS with a C++ wrapper for tasks

LHelge committed:
Merge branch 'master' into FreeRTOS

LHelge committed:
Tested helper functions for reading multiple byte values from I2C

abbeman commented on FW/FreeRTOS/Task.cpp:
```diff
@@ -0,0 +1,71 @@
+    Task::Task(const char *name, const uint16_t stackSize, const UBaseType_t priority) {
+        for(uint32_t i = 0; i < configMAX_TASK_NAME_LEN; i++) {
+            m_name[i] = name[i];
+        }
```
If name is less than ```configMAX_TASK_NAME_LEN``` this will read from outside it. If name is longer, m_name won't be null terminated. I suggest:

```c++
strncpy(m_name, name, configMAX_TASK_NAME_LEN);
// strncpy() does not guarantee null-termination.
m_name[configMAX_TASK_NAME_LEN - 1] = '\0';
```

abbeman reviewed (changes requested):
Looks really good, I like the Task class design. Very similar to ideas I had around a pthread class wrapper... =)

LHelge committed:
Resolved bug i Task constructor.

LHelge committed:
Cleaned up Task class and addod some comments

abbeman reviewed (approved):
Looks good. If you want it noted in docs, I think you can do @note in doxygen.

LHelge committed:
Wrapped FreeRTOS queues as well.

Replaced StateMAchine with a separate Task and the internal event queue with a FreeRTOS queue.

LHelge commented:
Some more FreeRTOS stuff before merging pull request :P

LHelge committed:
More FreeRTOS additions:
* Timers
* StreamBuffers (Not tested)
* MessageBuffers (Not tested)

Some changes to the app as well.

TODO: Put accelerometer driver in a task.

abbeman commented on FW/FreeRTOS/Queue.hpp:
```diff
@@ -0,0 +1,88 @@
+         * @brief Add item to queue, if queue is full, thread will be paused until there
+         *        is room or timeout elapses. Do not use this within an interrupt, use
+         *        EnqueueFromInterrupt() instead. Will return true if item successfully
+         *        was added to queue, or false if timed out
```
How about the ownership of item, is the data pointed to copied into the queue or does the queue take ownership of the data? Good to note in the doxygen, to avoid confusion and rummaging through the FreeRTOS docs. Same thing is good to note in the EnqueueFromInterrupt() and Dequeue() methods.

LHelge replied:
All FreeRTOS data synchronization structures like queues and buffers copy data in and out. This should be perfectly safe.

abbeman replied:
No I don't mean there's something wrong, I just meant it might be good to note in the doxygen comment that the data pointed to will be copied. So FreeRTOS noobs like me don't need to go searching for the FreeRTOS docs to see if the message passing takes ownership of the data or simply copies it when we want to pass a message. But maybe that's universally understood when the code can run without an allocator.

abbeman commented on FW/FreeRTOS/Queue.hpp:
```diff
@@ -0,0 +1,88 @@
+         * @brief Number of milliseconds for no timeout (defined as 24 H)
+         * 
+         */
+        static constexpr uint32_t MaxTimeout = pdMS_TO_TICKS(86400000);
```
I'd vote for following the convention of consts in all-caps, i.e. MAX_TIMEOUT etc.

LHelge replied:
Playing around with this... I actually think it's more readable this way. 

Heard a talk recently about writing code that you more or less can read as a English text. Avoiding abbreviations and "shouty" all-caps defines. One of the reasons being that you involve more parts of your brain while reading the code, not only the programming part but also the lingual. And that allows you to more quickly understand the code... Not sure what I think yet, but it's funny playing around with in a  project like this.

abbeman replied:
OK, hadn't really thought of it like that. I'm very used to the constant shouting (pun intended). Perhaps it's a foregone conclusion that it's good to separate variables from constants in the naming scheme, not sure though. Also, I guess they still are if you snake-case variable names...

abbeman commented on FW/FreeRTOS/MessageBuffer.hpp:
```diff
@@ -0,0 +1,78 @@
+#include "message_buffer.h"
+
+
+namespace FreeRTOS
```
Usually you don't indent code that is in a namespace, similarly to how you don't indent code inside an include guard, since it just adds an indent tax of 1 to everything.

abbeman commented on FW/FreeRTOS/Task.cpp:
```diff
@@ -0,0 +1,58 @@
+#include "task.h"
+
+
+extern "C" {
```
I don't think you usually indent code within an extern "C" block either. Come to think of it, maybe this file would make sense to make a .c file?

abbeman commented on FW/FreeRTOS/Task.hpp:
```diff
@@ -0,0 +1,146 @@
+        Task(const char *name, UBaseType_t priority) {
+            // TODO: assert(stackSize % sizeof(StackType_t));
+
+            strncpy(m_name, name, configMAX_TASK_NAME_LEN);
```
As an aside, I realized later you could also do:
```c++
snprintf(m_name, configMAX_TASK_NAME_LEN, "%s", name);
```
since snprintf() guarantees null termination, but nevermind now, the code as written is perfectly fine =)

abbeman commented on FW/FreeRTOS/Task.hpp:
```diff
@@ -0,0 +1,146 @@
+
+
+        // Four prio-levels should be enough for now
+        static constexpr UBaseType_t IdlePriority   = tskIDLE_PRIORITY;
```
```c++
enum TaskPriority : UBaseType_t {
    IDLE_PRIORITY = tskIDLE_PRIORITY,
    LOW_PRIORITY,
    MEDIUM_PRIORITY,
    HIGH_PRIORITY
};
```
?

abbeman commented on FW/FreeRTOS/Task.hpp:
```diff
@@ -0,0 +1,146 @@
+         * 
+         * @param milliseconds 
+         */
+        void InitPeriod(uint32_t milliseconds)  {
```
I'd suggest StartPeriod(), as init sounds like the function will set things up in preparation for use, rather than actually start the period. Start and end are good antonyms.

Just to understand, is this used to have a minimum time between periodic events, e.g. to make sure you don't invoke a method more often than, say once every 5 seconds?

LHelge replied:
One of the more common constructs I've seen in RTOS-work is to have an endless loop running at a fixed period independent on how much time the execution takes inside the loop. By having a simple delay, the period will be execution time + delay, which could be non-deterministic. This is a helper to realize such a loop. `InitPeriod(time)` before the loop, and then end the loop with `EndPeriod()` to delay `time - execution time`.

abbeman replied:
OK, I understand. I've done similar things with a timeout calculated in advance, doing some work and then waiting for events until the pre-calculated timeout hits. I just wasn't familiar with the start/end period way of looking at it.

LHelge committed:
Changes after comments

abbeman commented on FW/FreeRTOS/Task.hpp:
```diff
@@ -0,0 +1,146 @@
+         * @param instance Instance of this task
+         */
+        static void TaskFunction(void *instance)  {
+            Task *task = (Task*)instance;
```
I think google's style guide makes a pretty good case for not using C-style casts:
https://google.github.io/styleguide/cppguide.html#Casting
If I'm not mistaken, this would be appropriate here:
```c++
Task *task = static_cast<Task*>(instance);
```

abbeman commented on FW/FreeRTOS/Task.c:
```diff
@@ -0,0 +1,55 @@
+        configMINIMAL_STACK_SIZE is specified in words, not bytes. */
+    *pulIdleTaskStackSize = configMINIMAL_STACK_SIZE;
+}
+/*��������������������*/
```
What happened here? =D
